meteor/server/publications/packageManager/expectedPackages/generate.ts

```diff
@@ -0,0 +1,250 @@
+						allDeviceIds.add(deviceId)
+					}
+				}
+
+				for (const deviceId of allDeviceIds) {
+					// Filter, keep only the routed mappings for this device:
+					if (filterPlayoutDeviceIds && !filterPlayoutDeviceIds.includes(deviceId)) continue
+
+					const routedPackage = generateExpectedPackageForDevice(
+						studio,
+						{
+							...expectedPackage,
+							_id: `${pieceInstanceId}_${sanitisedPackageId}`,
+							rundownId: pieceInstanceDoc.rundownId,
+							contentVersionHash: getContentVersionHash(expectedPackage),
+						},
+						deviceId,
+						pieceInstanceId,
+						Priorities.OTHER // low priority
+					)
+
+					updatedDocIds.add(routedPackage._id)
+					collection.insert(routedPackage)
+				}
+			})
+		}
+	}
+
+	// Remove all documents for an ExpectedPackage that was regenerated, and no update was issues
+	collection.remove(
+		(doc) => !!doc.pieceInstanceId && updatedDocIds.has(doc._id) && !regenerateIds.has(doc.pieceInstanceId)
+	)
+}
+
+enum Priorities {
+	// Lower priorities are done first
+
+	/** Highest priority */
+	PLAYOUT_CURRENT = 0,
+	/** Second-to-highest priority */
+	PLAYOUT_NEXT = 1,
+	OTHER = 9,
+}
+
+function generateExpectedPackageForDevice(
+	studio: Pick<StudioLight, '_id' | 'packageContainers' | 'previewContainerIds' | 'thumbnailContainerIds'>,
+	expectedPackage: PackageManagerExpectedPackageBase,
+	deviceId: PeripheralDeviceId,
+	pieceInstanceId: PieceInstanceId | null,
+	priority: Priorities
+): PackageManagerExpectedPackage {
+	// Lookup Package sources:
+	const combinedSources: PackageContainerOnPackage[] = []
+
+	for (const packageSource of expectedPackage.sources) {
+		const lookedUpSource = studio.packageContainers[packageSource.containerId]
+		if (lookedUpSource) {
+			// We're going to combine the accessor attributes set on the Package with the ones defined on the source
+			const combinedSource: PackageContainerOnPackage = {
+				...omit(clone(lookedUpSource.container), 'accessors'),
+				accessors: {},
+				containerId: packageSource.containerId,
+			}
+
+			/** Array of both the accessors of the expected package and the source */
+			const accessorIds = _.uniq(
+				Object.keys(lookedUpSource.container.accessors).concat(Object.keys(packageSource.accessors || {}))
+			)
+
+			for (const accessorId of accessorIds) {
+				const sourceAccessor: Accessor.Any | undefined = lookedUpSource.container.accessors[accessorId]
+
+				const packageAccessor: AccessorOnPackage.Any | undefined = packageSource.accessors?.[accessorId]
+
+				if (packageAccessor && sourceAccessor && packageAccessor.type === sourceAccessor.type) {
+					combinedSource.accessors[accessorId] = deepExtend({}, sourceAccessor, packageAccessor)
+				} else if (packageAccessor) {
+					combinedSource.accessors[accessorId] = clone<AccessorOnPackage.Any>(packageAccessor)
+				} else if (sourceAccessor) {
+					combinedSource.accessors[accessorId] = clone<Accessor.Any>(sourceAccessor) as AccessorOnPackage.Any
+				}
```

Comment:
Maybe this could be refactored to a separate method, to lower the complexity a bit?